src/datasets/load.py

```diff
@@ -2205,7 +2231,17 @@ def load_dataset_builder(
             error_msg += f'\nFor example `data_files={{"train": "path/to/data/train/*.{example_extensions[0]}"}}`'
         raise ValueError(error_msg)
 
+    hash = dataset_module.hash
     builder_cls = get_dataset_builder_class(dataset_module, dataset_name=dataset_name)
+    if len(builder_cls.builder_configs) > 1:
+        builder_config = builder_cls.builder_configs.get(config_name or builder_cls.DEFAULT_CONFIG_NAME)
+    elif len(builder_cls.builder_configs) == 1:
+        builder_config = builder_cls.BUILDER_CONFIGS[0]
+    else:
+        builder_config = None
+    if builder_config and builder_config.data_files is not None:
+        builder_config._resolve_data_files(base_path=builder_kwargs["base_path"], download_config=download_config)
+        hash = update_hash_for_cache(hash, data_files=builder_config.data_files)
```

Comment:
The data files are resolved here to get the `hash` in the cache directory tree `config/version/hash`

It also takes into account the metadata configs and the dataset info, and you can find the hash update in the dataset modules factories, like  `hash = update_hash_for_cache(hash, metadata_configs=metadata_configs, dataset_infos=dataset_infos)` in HubDatasetModuleFactoryWithoutScript